packages/message-relayer/src/service.ts

```diff
@@ -1,188 +1,150 @@
-   * Logger to transport logs. Defaults to STDOUT.
-   */
-  logger?: Logger
-
-  /**
-   * Metrics object to use. Defaults to no metrics.
-   */
-  metrics?: Metrics
+interface MessageRelayerState {
+  wallet: Signer
+  messenger: CrossChainMessenger
+  highestCheckedL2Tx: number
+  highestKnownL2Tx: number
 }
 
-export class MessageRelayerService extends BaseService<MessageRelayerOptions> {
-  constructor(options: MessageRelayerOptions) {
-    super('Message_Relayer', options, {
-      relayGasLimit: {
-        default: 4_000_000,
-      },
-      fromL2TransactionIndex: {
-        default: 0,
-      },
-      pollingInterval: {
-        default: 5000,
-      },
-      getLogsInterval: {
-        default: 2000,
+export class MessageRelayerService extends BaseServiceV2<
+  MessageRelayerOptions,
+  MessageRelayerState
+> {
+  constructor(options?: Partial<MessageRelayerOptions>) {
+    super({
+      name: 'Message_Relayer',
+      options,
+      optionsSpec: {
+        l1RpcProvider: validators.provider({
+          desc: 'Provider for interacting with L1.',
+        }),
+        l2RpcProvider: validators.provider({
+          desc: 'Provider for interacting with L2.',
+        }),
+        l1Wallet: validators.wallet({
+          desc: 'Wallet used to interact with L1.',
+        }),
+        fromL2TransactionIndex: validators.num({
+          desc: 'Index of the first L2 transaction to start processing from',
+          default: 0,
+        }),
       },
     })
   }
 
-  private state: {
-    messenger: CrossChainMessenger
-    highestCheckedL2Tx: number
-  } = {} as any
-
-  protected async _init(): Promise<void> {
-    this.logger.info('Initializing message relayer', {
-      relayGasLimit: this.options.relayGasLimit,
-      fromL2TransactionIndex: this.options.fromL2TransactionIndex,
-      pollingInterval: this.options.pollingInterval,
-      getLogsInterval: this.options.getLogsInterval,
-    })
+  protected async init(): Promise<void> {
+    this.state.wallet = this.options.l1Wallet.connect(
+      this.options.l1RpcProvider
+    )
 
-    const l1Network = await this.options.l1Wallet.provider.getNetwork()
+    const l1Network = await this.state.wallet.provider.getNetwork()
     const l1ChainId = l1Network.chainId
     this.state.messenger = new CrossChainMessenger({
-      l1SignerOrProvider: this.options.l1Wallet,
+      l1SignerOrProvider: this.state.wallet,
       l2SignerOrProvider: this.options.l2RpcProvider,
       l1ChainId,
     })
 
     this.state.highestCheckedL2Tx = this.options.fromL2TransactionIndex || 1
+    this.state.highestKnownL2Tx =
+      await this.state.messenger.l2Provider.getBlockNumber()
   }
 
-  protected async _start(): Promise<void> {
-    while (this.running) {
-      await sleep(this.options.pollingInterval)
+  protected async main(): Promise<void> {
+    // If we're already at the tip, then update the latest tip and loop again.
+    if (this.state.highestCheckedL2Tx > this.state.highestKnownL2Tx) {
```

Comment:
I reimplemented this to take advantage of the fact that the base service now loops around main() by default, meaning we don't need to do our own looping logic.